### ISSUE TITLE
Crash when calling `addDevice()` from background thread 

### ISSUE THREAD

**Issue by vjard33** [GetStream/stream-chat-swift#2384]
## What did you do?
Try to add a device after connectUser, sometimes it crashes (this is the biggest crash we have on production on our app: 256 crashes).
ChatClient.shared.currentUserController().addDevice(.firebase(token: token, providerName: providerName))

## What did you expect to happen?
Not crash. ChatClient.shared.currentUserController().addDevice should addDevice if the token doesn't exist, if it exists: nothing.
```
        // Connect the client with the token provider.
        ChatClient.shared.connectUser(userInfo: userInfo, tokenProvider: tokenProvider) { error in
            if error == nil {
                // Stream Chat push notifications.
                if ChatClient.shared.currentUserId != nil {
                    if let fcmToken = Messaging.messaging().fcmToken {
                        ChatClient.shared.currentUserController().addDevice(.firebase(token: fcmToken, providerName: ChatClient.providerName))
                    }
                }

                successCompletion?()
            } else {
                failCompletion?(error)
            }
        }
```

## What happened instead?
Crash 
<img width="1424" alt="Capture d’écran 2022-11-22 à 21 21 50" src="https://user-images.githubusercontent.com/99725306/203414485-6c4a0812-953c-40f0-8b5e-787432b52f6d.png">

## GetStream Environment
**GetStream Chat version:** 4.22.0
**GetStream Chat frameworks:** StreamChat, StreamChatUI
**iOS version:** iOS 16.0.0, iOS 16.1.0, iOS 16.1.1
**Swift version:**5.7
**Xcode version:**14.1
**Device:**iPhone XS to iPhone 14 (seems to be more reproducible on iPhone 11)

## Additional context
I can't upgrade to 4.24.0 because I've weird behavior with TokenProvider after some logout/connectUser (completion called twice or more..)

**Comment by nuno-vieira** [comment]
Hi @vjard33,

When you are logging out a user, do you call `logout()`? Can you see in the analytics of your app the user journey and check if he previously logged out and then logged in? 

More details here: https://getstream.io/chat/docs/sdk/ios/uikit/getting-started/#disconnect--logout

Best,
Nuno

**Comment by nuno-vieira** [comment]
@vjard33 The docs also mention to add the device on the register for push notifications call, here: https://getstream.io/chat/docs/sdk/ios/client/push-notifications/

Any reason why you are doing it after the connectUser call?

**Comment by polqf** [comment]
> I can't upgrade to 4.24.0 because I've weird behavior with TokenProvider after some logout/connectUser (completion called twice or more..)

Side note, @vjard33, this has been fixed in our latest release  https://github.com/GetStream/stream-chat-swift/releases/tag/4.24.1

**Comment by nuno-vieira** [comment]
Hi @vjard33,

Let us know if this crash still happens on 4.24.1. We believe that it also fixes this problem. 

Best,
Nuno

**Comment by vjard33** [comment]
@nuno-vieira @polqf 
Waiting to see if the crash doesn't occur with the 4.24.1 (deployed to production this morning).
But this afternoon I saw a bug on `ChatClient.shared.currentUserController().currentUser` after a `connectUser()` which totaly freezes the app `[NSRecursiveLock lock]` called on your side after `Cached.reset()` 
You can see screenshots.

<img width="1435" alt="Capture d’écran 2022-11-28 à 15 22 22" src="https://user-images.githubusercontent.com/99725306/204302541-996deff4-381e-45da-b149-3ad05d10a76f.png">
<img width="1440" alt="Capture d’écran 2022-11-28 à 15 25 27" src="https://user-images.githubusercontent.com/99725306/204302566-93a70478-4885-4671-ac93-b2f49fe00713.png">

**Comment by nuno-vieira** [comment]
Hi @vjard33,

This is actually a known limitation/issue currently in our SDK. Currently, you should only use the Core SDK controllers in the main thread. So `addDevice()` or `currentUser` from the `ChatCurrentUserController` should be called in the main thread. Because you call these inside the `connectUser` completion block, you should first dispatch it to the main thread. This is an ongoing issue we have and we plan to fix it. For now, you can dispatch it to the main thread to workaround it.

**Comment by nuno-vieira** [comment]
@vjard33, I've updated the name of the issue, so that we can track this limitation of the SDK, and so that it is visible for everyone.

**Comment by nuno-vieira** [comment]
@vjard33 If you make sure you call the controllers from the main thread, you should not have any issues. Anyway, the addDevice call is actually an easy fix, so we will be able to fix it soon.

**Comment by vjard33** [comment]
@nuno-vieira Any news? Please fix thread for addDevice() AND currentUser from the ChatCurrentUserController. We use now the main thread, but we shouldn't do this on our side.

**Comment by vjard33** [comment]
@martinmitrevski Thanks for the update. On which week's day will you release the new SDK version please?

**Comment by martinmitrevski** [comment]
@vjard33 it will be most likely Wednesday or Thursday this week.

**Comment by vjard33** [comment]
@martinmitrevski @nuno-vieira @polqf Hello team, can you confirm me that I can call addDevice() AND currentUser without DispatchQueue.main.async with the future SDK 4.25.0? Thanks!

**Comment by nuno-vieira** [comment]
@vjard33 Yes, I can confirm, you can use `currentUserController.currentUser` from a background thread 👍

**Comment by polqf** [comment]
Hi! This has been fixed in our latest release [4.25.0](https://github.com/GetStream/stream-chat-swift/releases/tag/4.25.0)
Closing this issue for now, feel free to reopen it if you encounter any issues.
